kubernetes.go

```diff
@@ -98,29 +91,127 @@ func (k *Kubernetes) GetNode(name string) (node *apiv1.Node, err error) {
+// filterOutPodByOwnerReferenceKind filter out a list of pods by its owner references kind
+func filterOutPodByOwnerReferenceKind(podList []*apiv1.Pod, kind string) (output []*apiv1.Pod) {
+	for _, pod := range podList {
+		for _, ownerReference := range pod.Metadata.OwnerReferences {
+			if *ownerReference.Kind != kind {
+				output = append(output, pod)
+			}
+		}
+	}
+
+	fmt.Printf("\n\n%v\n\n", output)
+
 	return
 }
 
-// WatchNodes watch for updated preemptible node from a given node pool
-func (k *Kubernetes) WatchPreemptibleNodes() (watcher *k8s.CoreV1NodeWatcher, err error) {
-	watcher, err = k.Client.CoreV1().WatchNodes(context.Background(), k.PreemptibleNodeLabel())
+// DrainNode delete every pods from a given node and wait that all pods are removed before it succeed
+// it make sure we don't select DaemonSet as, they are not subject to unschedulable state
+func (k *Kubernetes) DrainNode(name string, drainTimeout int) (err error) {
+	// Select all pods sitting on the node except the one from kube-system
+	fieldSelector := k8s.QueryParam("fieldSelector", "spec.nodeName="+name+",metadata.namespace!=kube-system")
+
+	podList, err := k.Client.CoreV1().ListPods(context.Background(), k8s.AllNamespaces, fieldSelector)
+
+	// Filter out DaemonSet from the list of pods
+	filteredPodList := filterOutPodByOwnerReferenceKind(podList.Items, "DaemonSet")
+
+	if err != nil {
+		return
+	}
+
+	Logger.Info().
+		Str("host", name).
+		Msgf("%d pod(s) found", len(filteredPodList))
+
+	for _, pod := range filteredPodList {
+		Logger.Info().
+			Str("host", name).
+			Msgf("Deleting pod %s", *pod.Metadata.Name)
+
+		err = k.Client.CoreV1().DeletePod(context.Background(), *pod.Metadata.Name, *pod.Metadata.Namespace)
+
+		if err != nil {
+			Logger.Error().
+				Err(err).
+				Str("host", name).
+				Msgf("Error draining pod %s", *pod.Metadata.Name)
+			continue
+		}
+	}
+
+	doneDraining := make(chan bool)
+
+	// Wait until all pods are deleted
+	go func() {
+		for {
+			sleepTime := ApplyJitter(10)
+			sleepDuration := time.Duration(sleepTime) * time.Second
+			pendingPodList, err := k.Client.CoreV1().ListPods(context.Background(), k8s.AllNamespaces, fieldSelector)
+
+			if err != nil {
+				Logger.Error().
+					Err(err).
+					Str("host", name).
+					Msgf("Error getting list of pods, sleeping %ds", sleepTime)
+
+				time.Sleep(sleepDuration)
+				continue
+			}
+
+			// Filter out DaemonSet from the list of pods
+			filteredPendingPodList := filterOutPodByOwnerReferenceKind(pendingPodList.Items, "DaemonSet")
+			podsPending := len(filteredPendingPodList)
+
+			if podsPending == 0 {
+				doneDraining <- true
+				return
+			}
+
+			Logger.Info().
+				Str("host", name).
+				Msgf("%d pod(s) pending deletion, sleeping %ds", podsPending, sleepTime)
+
+			time.Sleep(sleepDuration)
+		}
+	}()
+
+	select {
+	case <-doneDraining:
+		break
+	case <-time.After(time.Duration(drainTimeout) * time.Second):
+		Logger.Warn().
```

Comment:
you can use `log` instead of `Logger` for the global logger. in `main.go` you can then initialize it with something like

```
	// set some default fields added to all logs
	log := zerolog.New(os.Stdout).With().
		Timestamp().
		Str("app", "estafette-preemptible-killer").
		Str("version", version).
		Logger()
```

to make sure it has some default fields it adds to all logs.